src/System.Private.Uri/src/System/UriExt.cs

```diff
@@ -561,10 +561,19 @@ public static string UnescapeDataString(string stringToUnescape)
 
                     UnescapeMode unescapeMode = UnescapeMode.Unescape | UnescapeMode.UnescapeAll;
                     position = 0;
-                    char[] dest = new char[stringToUnescape.Length];
-                    dest = UriHelper.UnescapeString(stringToUnescape, 0, stringToUnescape.Length, dest, ref position,
+                    ValueStringBuilder pooledArray = new ValueStringBuilder(stringToUnescape.Length, true);
+                    UriHelper.UnescapeString(stringToUnescape, 0, stringToUnescape.Length, ref pooledArray, ref position,
                         c_DummyChar, c_DummyChar, c_DummyChar, unescapeMode, null, false);
-                    return new string(dest, 0, position);
+
+                    if (pooledArray.IsSameString(pStr, 0, position))
```

Comment:
You don't need a custom compare operation for this.  You can just do:
```C#
if (pooledArray.AsSpan().SequenceEqual(new ReadOnlySpan<char>(pStr, position)))
```
which will be less code, safer, and generally more efficient.